### PR TITLE
support local creds probing

### PR DESCRIPTION
**Pull request by yugangw-msft**
For #118. 
The probe sequence:
1. Manager service identity
   only "system assigned identity" is probed, as user assigned identity would require extra arguments, and under that context users are better off using strong type credentials 
2. Environment variable `AZURE_CONNECTION_STRING` with sdk auth file content in json, which is proposed in the design review meeting in June.
    e.g. `{AZURE_CLIENT_ID='...', AZURE_CLIENT_SECRET='...', AZURE_TENANT_ID='...', AZURE_CLOUD_NAME='...'}`
    `AZURE_CLOUD_NAME` is optional. We default to public cloud
3. Individual environment variables for a service principal
    Like 2, rather using individual environment variables
4. Azure Authentication File (originating from java sdk)
5. Azure CLI credentials

Subscription id will be optional as the chainer will find them
Per suggestion, the example code includes both management plane and data plane client codes

### CODE REVIEW

**Comment by yugangw-msft** [comment]
//cc: @lmazuel  @jianghaolu

**Comment by yugangw-msft** [comment]
cc: @jianghaolu  @lmazuel  @jhendrixMSFT @shahabhijeet @joshgav

File: msrestazure/azure_local_creds_prober.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        super(CLICredentials, self).__init__(None)
+        uname = platform.uname()
+        # python 2, `platform.uname()` returns: tuple(system, node, release, version, machine, processor)
+        platform_name = getattr(uname, 'system', None) or uname[0]
+        platform_name = platform_name.lower()
+        if platform_name == 'windows':
+            program_files_folder = os.environ.get('ProgramFiles(x86)') or os.environ.get('ProgramFiles')
+            probing_paths = [os.path.join(program_files_folder, 'Microsoft SDKs', 'Azure', 'CLI2', 'wbin', 'az.cmd')]
+        else:
+            probing_paths = ['/usr/bin/az', '/usr/local/bin/az']
```
Comment (lmazuel) [inline msrestazure/azure_local_creds_prober.py]: This won't work if CLI was installed by pip. You could subprocess "which az", or us "shutil.which('az')" starting Python 3
https://docs.python.org/3/library/shutil.html?highlight=.which#shutil.which

File: msrestazure/azure_local_creds_prober.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        stdout, stderr = p.communicate()
+        p.wait()
+        if stderr:
+            raise ValueError('Retrieving acccess token failed: ' + stderr)
+        info = json.loads(stdout)
+        self.scheme, self.token = info['tokenType'], {'access_token': info['accessToken']}
+
+    def signed_session(self, session=None):
+        # Token cache is handled by the VM extension, call each time to avoid expiration
+        self.set_token()
```
Comment (lmazuel) [inline msrestazure/azure_local_creds_prober.py]: Should, the CLI be called each request to refresh the token?

File: tests/test_local_creds_prober.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# THE SOFTWARE.
+#
+#--------------------------------------------------------------------------
+
+
+from azure.mgmt.storage import StorageManagementClient
+from msrestazure.azure_local_creds_prober import AzureLocalCredentialProber
+
+prober = AzureLocalCredentialProber()
+client = StorageManagementClient(prober, prober.subscription_id)
```
Comment (lmazuel) [inline tests/test_local_creds_prober.py]: I would prefer to use the factory style I have right now:
```python
from azure.common.client_factory import get_client_from_cli_profile
from azure.mgmt.compute import ComputeManagementClient

client = get_client_from_cli_profile(ComputeManagementClient)
```

but something more generic like:

```python
from azure.common.client_factory import get_client
from azure.mgmt.compute import ComputeManagementClient

client = get_client(ComputeManagementClient)
```

File: msrestazure/azure_local_creds_prober.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import requests
+from msrest.authentication import BasicTokenAuthentication
+from msrestazure.azure_active_directory import MSIAuthentication, ServicePrincipalCredentials
+
+_LOGGER = logging.getLogger(__name__)
+
+class AzureLocalCredentialProber(object):
+    '''
+    Probing logics:
+    1. Managed service identity
```
Comment (johanste) [inline msrestazure/azure_local_creds_prober.py]: The progression should be from more specific to less specific. This is the way I'd define the specificity of the different sources of information:

```
Environment variables > local file > local endpoint available. 
```
Following this, does this: 2, 3, 1, 4 make sense?

File: msrestazure/azure_local_creds_prober.py (hunk 1 above)
Comment (johanste) [reply msrestazure/azure_local_creds_prober.py]: Can we use the presence of a ${HOME}/.azure file as the hint to try the CLI creds?

Comment (johanste) [reply msrestazure/azure_local_creds_prober.py]: ...or, alternatively, use the ${PATH} environment variable...

**Comment by joshgav** [comment]
Explicit environment variables should come before managed identity like @johanste says cause we should prefer explicit intent. That's what the specs say too :).

Rather than a global "Prober", could each credentials type include a static `Probe()` method which returns a boolean? That would grant more flexibility to the provider and allow us to add new providers and probes in the future.

Can we document the names of keys in the connection string and names of environment variables? We should use the same names in all SDKs.

**Comment by joshgav** [comment]
We will need to get tokens for resources other than AzureRM, e.g. Storage and Key Vault. Can the constructor be parameterized to take a resource URL and default to "https://management.azure.com"?

File: msrestazure/azure_local_creds_prober.py (hunk 2 above)
Comment (yugangw-msft) [reply msrestazure/azure_local_creds_prober.py]: I assume you meant sub-shelling is kind of slow? Yes, I can optimize by leveraging the `expiresOn` information returned from the `get-access-token` command

File: msrestazure/azure_local_creds_prober.py (hunk 4 above)
Comment (yugangw-msft) [reply msrestazure/azure_local_creds_prober.py]: I started with the most secure option, then follow up with SDK home made ones. But you have a point, so I will adjust. Also it is unlikely people would set env variable in a VM with identity

**Comment by yugangw-msft** [comment]
> We will need to get tokens for resources other than AzureRM, e.g. Storage and Key Vault

Each specific SDK has the resource uri defined inside, e.g. [go-sdk](https://github.com/Azure/azure-sdk-for-go/blob/master/services/keyvault/mgmt/2015-06-01/keyvault/client.go#L29) , so this is not a  concern.
Since we are at it, let us see how to get sovereign clouds work. In Python's autorest runtime, those endpoints are pre-defined as [a set of instances](https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/azure_cloud.py#L131).  I believe Java does too (cc: @jianghaolu).  If other language SDKs have the similar supports, then the probing can default to public azure while client app can override with an enum.

**Comment by yugangw-msft** [comment]
Hi folks, i have addressed all feedback, please take another look.
The sample code wrapped in the unit test works for both storage management plane and data plane. Note the new storage data plan involves a highly unique requirement on client applications on how to supply the access token.  I will cross check with my CLI colleague for whether or not I am doing the most optimal way. 
The method now exposes the `resource` keyword arguments with the defaulting value on arm.

File: msrestazure/azure_local_creds_prober.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+
+class ConnectionStrEnvProber(CredsProber):
+    '''
+    Detect environment variable AZURE_CONNECTION_STRING
+    '''
+    def probe(self):
+        creds = None
+        if self.enabled and os.environ.get('AZURE_CONNECTION_STRING'):
+            auth_info = json.loads(os.environ.get('AZURE_CONNECTION_STRING'))
```
Comment (lmazuel) [inline msrestazure/azure_local_creds_prober.py]: Is the connection string JSON valid? I thought it was something like:
AZURE_CONNECTION_STRING="client_id=1234;client_secret=2345;tenant_id=3456;subscription_id=4567;"

File: msrestazure/azure_local_creds_prober.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            creds = ServicePrincipalCredentials(client_id=auth_info['clientId'],
+                                                secret=auth_info['clientSecret'],
+                                                tennt_id=auth_info['tenantId'],
+                                                resource=resource,
+                                                cloud_environment=cloud_environment)
+            return creds
+        return None
+
+    def probe_subscription(self, creds):
+        return json.loads(os.environ.get('AZURE_CONNECTION_STRING'))['subscriptionId']
```
Comment (lmazuel) [inline msrestazure/azure_local_creds_prober.py]: Same question about JSON, and I thought too it should be `subscription_id`

File: msrestazure/azure_local_creds_prober.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    def probe(self):
+        creds = None
+        if self.enabled and os.environ.get('AZURE_CONNECTION_STRING'):
+            auth_info = json.loads(os.environ.get('AZURE_CONNECTION_STRING'))
+            if auth_info.get('cloudName'):
+                cloud_environment = self.find_cloud_env(auth_info.get('cloudName'))
+            else:
+                cloud_environment = AZURE_PUBLIC_CLOUD
+            resource = self.resource or cloud_environment.endpoints.management
+            creds = ServicePrincipalCredentials(client_id=auth_info['clientId'],
```
Comment (lmazuel) [inline msrestazure/azure_local_creds_prober.py]: client_id?

File: msrestazure/azure_local_creds_prober.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    Probing logics:
+    1. AZURE_CONN_STR, with SDK auth code file content in json.
+        https://github.com/Azure/azure-sdk-for-java/wiki/Authentication
+    2. Individual environment variables to estabslish a service principal's creds
+        https://github.com/Azure/azure-sdk-for-go
+    3. Managed service identity
+        a. app service
+        b. virtual machine
+    4. Azure CLI, through "az account get-access-token"
+    '''
```
Comment (lmazuel) [inline msrestazure/azure_local_creds_prober.py]: Add the authentication file:
https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python#mgmt-auth-file

File: msrestazure/azure_local_creds_prober.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    for prober in probers:
+        creds = prober.probe()
+        if creds:
+            return creds, prober
+    raise ValueError('No credential was detected from the local machine')
+
+
+def get_creds_through_local_probing(**kwargs):
+    '''
+    Probing logics:
```
Comment (lmazuel) [inline msrestazure/azure_local_creds_prober.py]: Should we have the concept of critical failure? If we recognize some env variables and something is wrong, is this enough to fail completly and not fallback?
@johanste @joshgav

Comment (johanste) [reply msrestazure/azure_local_creds_prober.py]: Yes.

**Comment by lmazuel** [comment]
Will be done in azure-identity but not msrestazure

**Pull request closed without merging**